src/main/scala/io/iohk/ethereum/domain/branch/BestBlockchainBranch.scala

```diff
@@ -0,0 +1,37 @@
+package io.iohk.ethereum.domain.branch
+import akka.util.ByteString
+
+import io.iohk.ethereum.db.storage.BlockNumberMappingStorage
+import io.iohk.ethereum.domain.Block
+import io.iohk.ethereum.domain.BlockHeader
+import io.iohk.ethereum.domain.BlockchainReader
+
+/** A Branch instance which only works for the best canonical branch. As this branch
+  * currently has specific indexes (particularly regarding accessing blocks by number),
+  * it will uses thoses to provide better performance.
+  */
+class BestBlockchainBranch(
+    tipBlockHeader: BlockHeader,
+    bestChainBlockNumberMappingStorage: BlockNumberMappingStorage,
+    blockchainReader: BlockchainReader
+) extends BlockchainBranch {
+
+  /* The following assumptions are made in this class :
+   *  - the whole branch exist in storage
+   *  - The various metadata and index are consistent
+   *
+   * If those assumptions are found to be false, then the application is in an inconsistent
+   * state and the class will throw.
+   * */
+
+  override def getBlockByNumber(number: BigInt): Option[Block] =
+    if (tipBlockHeader.number >= number && number >= 0) {
```

Comment:
This is some additionnal logic compared to what was existing. The rationale is that if we want a branch to be an immutable view, then I have to avoid fetching block "in the future" compared to the selected tip. 

This is still not perfect :
  - if the `number -> block hash` index  is rewritten, then the interface is not really immutable. The problem should not be apparent unless the instance is used for too long.
  - `getHashByBlockNumber` does not offer such guarantee currently because it would need to fetch the whole block to check, which kind of defeat the purpose of this function. But this function is not used in a lot of place so it might make sense to not expose it there or make it protected.